Add a document to explain how to install ROS and Gazebo together (#339)
gazebosim/docs: ros_installation.md

 * ***packages.ros.org***
   * ROS1 Noetic: Gz Citadel
   * ROS2 Foxy: Gz Citadel
   * ROS2 Galactic: Gz Edifice
   * ROS2 Humble: Gz Fortress
   * ROS2 Rolling: Gz Fortress (changing frequently)

 * ***packages.osrfoundation.org***
   * Gz Citadel
   * Gz Edifice
   * Gz Fortress
   * Gz Garden

This means that including the osrfoundation repository is not strictly needed
to get the Gazebo binary package in Ubuntu. It can be installed from the ROS
repository.

## Installing the default Gazebo version for a ROS distribution using binary installations

For users that need to run a specific version of ROS and want to use all
the gazebo ROS related packages out-of-the-box, this is the recommended
section:

### ROS2 Humble and ROS2 Rolling
ROS 2 Humble and ROS 2 Rolling support officially the Gazebo Fortress version.
For a fully-integrated ROS system, Fortress is the recommened version. The way
to proceed is just to use the ROS repository and do ***not*** use the
osrfoundation repository.

ROS packages `ros-humble-ros-gz` and `ros-rolling-ros-gz` will be available
, these packages automatically install the necessary dependencies for Gazebo Fortress.

> ### ROS 2 Galactic

Reviewer: End of life ?

```suggestion
### ROS 2 Galactic 
```

ROS 2 Galactic officially supports Gazebo Edifice

>  For a fully-integrated ROS system, Edifice is the recommened

Reviewer: ```suggestion
For a fully-integrated ROS system, Edifice is the recommened
```

   from osrfoundation.org repository
 * Follow the instructions to compile the ros_gz bridge from source in a colcon workspace
   * [ROS 2 Galactic](https://github.com/gazebosim/ros_gz/tree/galactic#from-source)
     * Be sure of using `export GZ_VERSION=fortress`
   * [ROS1 Noetic]()https://github.com/gazebosim/ros_gz/tree/noetic#from-source
     * Be sure of using `export GZ_VERSION=fortress`

### Gz Edifice
Gz Edifice can be also used with ROS2 Foxy, ROS2 Rolling and ROS1 Noetic, the [`ros_gz
bridge`](https://github.com/gazebosim/ros_gz) supports the compilation from
source.

 * Folow [the instruction to install gz-edifice](https://gazebosim.org/docs/edifice/install_ubuntu#binary-installation-on-ubuntu)
   from osrfoundation.org repository
 * Follow the instructions to compile the ros_gz bridge from source in a colcon workspace
   * [ROS 2 Foxy](https://github.com/gazebosim/ros_gz/tree/foxy#from-source)
     * Be sure of using `export GZ_VERSION=edifice`
   * [ROS 2 Rolling](https://github.com/gazebosim/ros_gz/tree/ros2#from-source)
     * Be sure of using `export GZ_VERSION=edifice`

>    * [ROS1 Noetic]()https://github.com/gazebosim/ros_gz/tree/noetic#from-source

Reviewer: ```suggestion
   * [ROS1 Noetic](https://github.com/gazebosim/ros_gz/tree/noetic#from-source)
```

# Install Gazebo and ROS

This document provides an overview about the options to use different versions

> of ROS in combination with different versions of Gz. It is recommended to read

Reviewer 2: nit: Gz -> Gazebo.
Here and below

Author: replaced all them [03cd75f](https://github.com/gazebosim/docs/pull/339/commits/03cd75fdd4c1b2f6fa0d0d8895f1168e67233205)

it before installing the [ros_gz bridge](https://github.com/gazebosim/ros_gz).

# Important! Simple analysis for a quick and correct decision

If you are planning on using a specific version of ROS and don't have a reason
to use a specific version of Gazebo, you should proceed with the default

> version of Gazebo available from the ROS repository when installing the ros_gz

Reviewer: ```suggestion
version of Gazebo available from the ROS repository when installing the `ros_gz`
```

bridge:

> ```
> apt-get install ros-${ROS_DISTRO}-ros-gz
> 
> ```

Reviewer: ```suggestion
```bash
sudo apt-get install ros-${ROS_DISTRO}-ros-gz
```
```

This command is going to install the Gazebo libraries corresponding to the
version detailed below in the section: "Using the default Gazebo version
for a ROS distribution using binary installations". The rest of this
document provides more information and other possible installations but
at this point ROS and Gz are installed and ready.

# Gazebo versions and ROS integration

> Gazebo is an independent project like boost, ogre or any other project used by

Reviewer: ```suggestion
Gazebo is an independent project like boost, Ogre or any other project used by
```

ROS. Usually, the latest version of Gazebo available at the beginning of every

> ROS release cycle (for example Gazebo Fortress for ROS2 Humble) is selected as

Reviewer 3: ```suggestion
ROS release cycle (for example Gazebo Fortress for ROS 2 Humble) is selected as
```

the official one to be fully integrated and supported and will be kept during
the whole life of the ROS distribution.

Note that all Gz libraries policy follows the [semantic
versioning](http://semver.org/) philosophy for ABI stability, in which all
versions that have the same major number (`gz-sim7_7.0.0`, `gz-sim7_7.1.0`,
`gz-sim7_7.0.1`, ...) are binary compatible and thus interchangeable when using
the same ROS distro.

## Installing Gazebo

### Gz simulator Ubuntu packages

> The easiest way of installing Gazebo is to use packages. There are two main

Reviewer 3: ```suggestion
The easiest way of installing Gazebo is to use binary packages. There are two main
```

repositories which host Gz simulator and Gz libraries: one is `packages.ros.org` and the
other is `packages.osrfoundation.org`. At the time of writing:

 * ***packages.ros.org***
   * ROS1 Noetic: Gz Citadel
   * ROS2 Foxy: Gz Citadel
   * ROS2 Galactic: Gz Edifice
   * ROS2 Humble: Gz Fortress
   * ROS2 Rolling: Gz Fortress (changing frequently)

 * ***packages.osrfoundation.org***
   * Gz Citadel
   * Gz Edifice
   * Gz Fortress
   * Gz Garden

This means that including the osrfoundation repository is not strictly needed
to get the Gazebo binary package in Ubuntu. It can be installed from the ROS
repository.

> ## Using the default Gazebo version for a ROS distribution using binary installations

Reviewer 3: ```suggestion
## Installing the default Gazebo version for a ROS distribution using binary installations
```

> For the users that need to run a specific version of ROS and want to use all

Reviewer 3: ```suggestion
For users that need to run a specific version of ROS and want to use all
```

the gazebo ROS related packages out-of-the-box, this is the recommended
section:

### ROS2 Humble and ROS2 Rolling

> ROS2 Humble and ROS2 Rolling support officially the Gazebo Fortress version.

Reviewer 3: ```suggestion
ROS 2 Humble and ROS 2 Rolling support officially the Gazebo Fortress version.
```

For a fully-integrated ROS system, Fortress is the recommened version. The way

> to proceed is just to use the ROS repository and do ***not*** use the

Reviewer 3: To clarify,  when you say "use" here you specifically mean "download and install binaries from" correct?

Author: "use the X repository" I meant ... setup keys and sources.lists for the X repository. So yes, at the end "download and install binaries from X repository".

osrfoundation repository.

ROS packages `ros-humble-ros-gz` and `ros-rolling-ros-gz` will be available

> pulling the necessary Gazebo Fortress dependencies.

Reviewer 3: I think what you mean here is:
```suggestion
, these packages automatically install the necessary dependencies for Gazebo Fortress.
```

> ### ROS2 Galactic

Reviewer 3: ```suggestion
### ROS 2 Galactic
```

> ROS2 Galactic supports officially the Gazebo Edifice

Reviewer 3: ```suggestion
ROS 2 Galactic officially supports Gazebo Edifice
```

> version. For a fully-integrated ROS system, Edifice is the recommened

Reviewer 3: ```suggestion
 For a fully-integrated ROS system, Edifice is the recommened
```

> version. The way to proceed is just to use the ROS repository and do ***not***

Reviewer 3: We have to be a bit clearer here.  I am not sure what use is supposed to mean.

use the osrfoundation repository.

ROS packages `ros-galactic-ros-gz` will be available pulling the necessary
Gazebo Edifice dependencies.

> ### ROS2 Foxy

Reviewer 3: ```suggestion
### ROS 2 Foxy
```

> ROS2 Foxy supports officially the Gazebo Citadel

Reviewer 3: ```suggestion
ROS 2 Foxy supports officially the Gazebo Citadel
```

version. For a fully-integrated ROS system, Citadel is the recommened
version. The way to proceed is just to use the ROS repository and do ***not***
use the osrfoundation repository.

ROS packages `ros-foxy-ros-gz` will be available pulling the necessary
Gazebo Citadel dependencies.

> ### ROS1 Noetic

Reviewer 3: ```suggestion
### ROS 1 Noetic
```

> ROS1 Noetic supports officially the Gazebo Citadel

Reviewer 3: ```suggestion
ROS Noetic officially supports Gazebo Citadel
```

version. The way to proceed is just to use the ROS repository and do ***not***
use the osrfoundation repository.

ROS packages `ros-noetic-ros-gz` will be available pulling the necessary
Gazebo Citadel dependencies.

## Using an specific Gazebo version with ROS 2
***Warning!:*** Using this option, you won't be able to use any ROS Ubuntu package
related to Gz libraries from ROS deb repository.

Selecting a different release of Gz than the one officially supported in each
ROS distribution will require that you manually compile the [`ros_gz
bridge`](https://github.com/gazebosim/ros_gz) from source. Same applies for

> other ROS packages using Gz libraries that could also require code changes in

Reviewer 2: "changes in ..." is there something missing there?

Author: Probably I meant an "if" [a828de5](https://github.com/gazebosim/docs/pull/339/commits/a828de56d9c787b7c2ac52daea497a708453edef)

other ROS packages using Gz libraries that could also require code changes in
they are not supporting the new Gz version chosen.

### Gz Garden

Gz Garden can be used with ROS2 Humble and ROS2 Rolling, the [`ros_gz
bridge`](https://github.com/gazebosim/ros_gz) supports the compilations from
source.

 * Folow [the instruction to install gz-garden](https://gazebosim.org/docs/garden/install_ubuntu#binary-installation-on-ubuntu)
   from osrfoundation.org repository

>  * Follow the instructions to compile the ros_gz bridge from source in a colcon workspace

Reviewer: ```suggestion
 * Follow the instructions to compile the `ros_gz` bridge from source in a colcon workspace
```

>  * Install [rosdep rules for Gazebo Garden](https://github.com/osrf/osrf-rosdep#installing-rosdep-rules-to-resolve-gazebo-garden-libraries)

Reviewer 2: wow, cool! I didn't know about this.

Author: brand new, from yesterday. Should help users with Garden installations in ROS.

   * [ROS2 Humble](https://github.com/gazebosim/ros_gz/tree/humble#from-source)
     * Be sure of using `export GZ_VERSION=humble`
   * [ROS2 Rolling](https://github.com/gazebosim/ros_gz/tree/ros2#from-source)

>      * Be sure of using `export GZ_VERSION=rolling`

Reviewer: This is wrong, maybe `ROS_VERSION`?

Author: ouch, [4dd64b9](https://github.com/gazebosim/docs/pull/339/commits/4dd64b91e508fd7fc8e856d6a719ad2e65ed3b98) . Good catch.

### Gz Fortress

> Gz Fortress can be also used with ROS2 Galactic and ROS1 Noetic, the [`ros_gz

Reviewer 3: ```suggestion
Gz Fortress can be also used with ROS 2 Galactic and ROS 1 Noetic, the [`ros_gz
```

bridge`](https://github.com/gazebosim/ros_gz) supports the compilation from
source.

 * Folow [the instruction to install gz-fortress](https://gazebosim.org/docs/fortress/install_ubuntu#binary-installation-on-ubuntu)
   from osrfoundation.org repository
 * Follow the instructions to compile the ros_gz bridge from source in a colcon workspace
   * [ROS2 Galactic](https://github.com/gazebosim/ros_gz/tree/galactic#from-source)
     * Be sure of using `export GZ_VERSION=fortress`
   * [ROS1 Noetic]()https://github.com/gazebosim/ros_gz/tree/noetic#from-source
     * Be sure of using `export GZ_VERSION=fortress`

### Gz Edifice
Gz Edifice can be also used with ROS2 Foxy, ROS2 Rolling and ROS1 Noetic, the [`ros_gz
bridge`](https://github.com/gazebosim/ros_gz) supports the compilation from
source.

 * Folow [the instruction to install gz-edifice](https://gazebosim.org/docs/edifice/install_ubuntu#binary-installation-on-ubuntu)
   from osrfoundation.org repository
 * Follow the instructions to compile the ros_gz bridge from source in a colcon workspace

>    * [ROS2 Foxy](https://github.com/gazebosim/ros_gz/tree/foxy#from-source)

Reviewer 3: ```suggestion
   * [ROS 2 Foxy](https://github.com/gazebosim/ros_gz/tree/foxy#from-source)
```

     * Be sure of using `export GZ_VERSION=foxy`

>    * [ROS2 Rolling](https://github.com/gazebosim/ros_gz/tree/ros2#from-source)

Reviewer 3: ```suggestion
   * [ROS 2 Rolling](https://github.com/gazebosim/ros_gz/tree/ros2#from-source)
```

>      * Be sure of using `export GZ_VERSION=rolling`

Reviewer: This is wrong GZ_VERSION should't be `rolling`

Author: right, all them should be Edifice [b37e02e](https://github.com/gazebosim/docs/pull/339/commits/b37e02eddb80bc21d66b3cff4f3c9589f3812272)

   * [ROS1 Noetic]()https://github.com/gazebosim/ros_gz/tree/noetic#from-source
     * Be sure of using `export GZ_VERSION=edifice`

## Using the latest Gz binary versions

The Gz team usually backport and release new versions of each of the supported

> Gz releases and libraries (i.e: bumping gz-sim 7.0.0 to gz-sim 7.1.0). These

Reviewer: ```suggestion
Gz releases and libraries (i.e: bumping `gz-sim 7.0.0` to gz-sim `7.1.0`). These
```

updates are hosted first in the osrfoundation repository. The ROS repository
syncs from the osrfoundation repository frequently but versions can be
different.

Getting the latest versions of the Gazebo libraries and simulator is as easy
as installing the [osrfoundation.org repository](https://gazebosim.org/docs/latest/install_ubuntu_src#install-dependencies)
together with the ROS repository. Updates should be fully compatible.

## FAQ

#### I am not using ROS at all, which version should I use?

If you don't need ROS support, the recommended version is the latest released
version that can be [installed using the osrfoundation repo](https://gazebosim.org/docs)
depending on your platform.

#### I want to use Gz Garden with ROS. Where are the packages?

Unfortunately there are no Gz Garden packages for any ROS distribution. Please
read how to compile it from source in this document under the section
"Using an specific Gazebo version with ROS2" >> "Gz Garden"

#### Where I can find the different features implemented on each Gazebo version?

> Some notes are regularly posted on the [Gazebo community

Reviewer 2: I think https://gazebosim.org/docs/garden/release-features would be a good link to put here.

Author: Indeed, that is the best option. [572f672](https://github.com/gazebosim/docs/pull/339/commits/572f672e2208102c39713e1be07577407418d124)
